File: node/examples/isolated-connections.ts
import { RPCClass } from '../src/decorators.js';
import { createRPCClient } from '../src/index.js';

interface TestService {
  ping(): Promise<string>;
  heavyComputation(seconds: number): Promise<number>;
  generateDataStream(): AsyncGenerator<string>;
}

@RPCClass
class TestServiceImpl implements TestService {
  async ping() {
    return `pong at ${new Date().toISOString()}`;
  }

  async heavyComputation(seconds: number) {
    console.log(`Starting heavy computation for ${seconds}s`);
    const start = performance.now();
    while (performance.now() - start < seconds * 1000) {
      await new Promise((resolve) => setTimeout(resolve, 1));
    }
    console.log('Heavy computation completed');
    return performance.now() - start;
  }

  async *generateDataStream() {
    for (let i = 0; i < 100; i++) {
      yield `Data chunk ${i} at ${new Date().toISOString()}`;
      await new Promise((resolve) => setTimeout(resolve, 10));
    }
  }
}

async function testIsolatedChannels() {
  console.log('=== Testing Isolated Channels ===\n');

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'test-client',
  });

  await client.connect();

  const regularChannel = await client.channel('regular-channel');

  const isolatedChannel = await client.channel('isolated-channel', {
    isolatedConnection: true,
  });

  regularChannel.on('message', (data) => {
    if (data.file) {
      data.file = `<${data.file.length} bytes>`;
    }

    console.log('Regular Channel:', data);
  });

  isolatedChannel.on('message', (data) => {
    if (data.file) {
      data.file = `<${data.file.length} bytes>`;
    }

    console.log('Isolated Channel:', data);
  });

  console.log('Sending heavy traffic on isolated channel...');
  const isolatedStart = performance.now();
  const promises: Promise<void>[] = [];
  const testBuffer = Buffer.alloc(10000).fill('x');
  for (let i = 0; i < 1000; i++) {
    promises.push(
      isolatedChannel.send({
        type: 'bulk',
        index: i,
        data: testBuffer,
      }),
    );
  }

  console.log('Testing regular channel responsiveness...');
  const regularStart = performance.now();
  await regularChannel.send({ type: 'ping', timestamp: performance.now() });
  const regularTime = performance.now() - regularStart;
  console.log(`Regular channel responded in ${regularTime.toFixed(1)}ms`);

  await Promise.all(promises);
  const isolatedTime = performance.now() - isolatedStart;
  const throughput = (1000 * 10) / (isolatedTime / 1000); // 10KB * 1000 messages
  console.log(`Heavy traffic completed in ${isolatedTime.toFixed(1)}ms (${(throughput / 1024).toFixed(1)} MB/s)\n`);

  await regularChannel.close();
  await isolatedChannel.close();
  await client.disconnect();
}

async function testIsolatedProxies() {
  console.log('=== Testing Isolated Proxies ===\n');

  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'test-server',
  });

  await server.connect();
  const unsub = await server.registerHandler('test', new TestServiceImpl());

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'test-client',
  });

  await client.connect();

  const regularProxy = client.createProxy<TestService>('test');

  const isolatedProxy = client.createProxy<TestService>('test', {
    isolatedConnection: true,
  });

  console.log('Starting heavy computation on isolated proxy...');
  const heavyStart = performance.now();
  const heavyPromise = isolatedProxy.proxy.heavyComputation(3);

  console.log('Testing regular proxy responsiveness...');
  const pingTimes: number[] = [];
  for (let i = 0; i < 5; i++) {
    const pingStart = performance.now();
    const result = await regularProxy.ping();
    const pingTime = performance.now() - pingStart;
    pingTimes.push(pingTime);
    console.log(`Regular proxy ping ${i + 1}: ${pingTime.toFixed(1)}ms - ${result}`);
    await new Promise((resolve) => setTimeout(resolve, 500));
  }
  const avgPingTime = pingTimes.reduce((a, b) => a + b, 0) / pingTimes.length;
  console.log(`Average ping time during heavy computation: ${avgPingTime.toFixed(1)}ms`);

  const computationTime = await heavyPromise;
  const totalHeavyTime = performance.now() - heavyStart;
  console.log(`Heavy computation completed in ${computationTime}ms (total: ${totalHeavyTime.toFixed(1)}ms)\n`);

  console.log('Testing streaming on isolated proxy...');
  const streamStart = performance.now();
  let count = 0;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  for await (const data of isolatedProxy.proxy.generateDataStream()) {
    count++;
    if (count % 20 === 0) {
      console.log(`Received ${count} chunks from isolated stream`);
    }
  }
  const streamTime = performance.now() - streamStart;
  console.log(`Stream completed: ${count} total chunks in ${streamTime.toFixed(1)}ms (${((count * 1000) / streamTime).toFixed(1)} chunks/s)\n`);

  await unsub();

  await isolatedProxy.close();
  await client.disconnect();
  await server.disconnect();

  console.log('Isolated proxies test completed\n');
}

async function testMixedWorkload() {
  console.log('=== Testing Mixed Workload ===\n');

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'mixed-client',
  });

  await client.connect();

  const chatChannel = await client.privateChannel('chat', 'partner', {
    isolatedConnection: true,
  });

  const dataChannel = await client.privateChannel('data-transfer', 'partner', {
    isolatedConnection: true,
  });

  console.log('Starting mixed workload...');

  const chatInterval = setInterval(async () => {
    await chatChannel.send({
      type: 'chat',
      message: 'Hello from chat',
      timestamp: new Date().toISOString(),
    });
  }, 1000);

  const dataStart = performance.now();
  const dataPromises: Promise<void>[] = [];
  const dataBuffer = Buffer.alloc(100 * 1024).fill('x');
  for (let i = 0; i < 100; i++) {
    dataPromises.push(
      dataChannel.send({
        type: 'data',
        chunk: i,
        payload: dataBuffer,
      }),
    );
  }

  let chatMessages = 0;
  let dataChunks = 0;

  chatChannel.on('message', () => chatMessages++);
  dataChannel.on('message', () => dataChunks++);

  await Promise.all(dataPromises);
  clearInterval(chatInterval);
  const dataTime = performance.now() - dataStart;
  const dataThroughput = (100 * 100) / (dataTime / 1000); // 100KB * 100 messages

  console.log(`Chat messages: ${chatMessages}`);
  console.log(`Data chunks: ${dataChunks}`);
  console.log(`Data transfer: 10MB in ${dataTime.toFixed(1)}ms (${(dataThroughput / 1024).toFixed(1)} MB/s)`);
  console.log('Mixed workload completed\n');

  await chatChannel.close();
  await dataChannel.close();
  await client.disconnect();
}

async function main() {
  const totalStart = performance.now();

  try {
    await testIsolatedChannels();
    await testIsolatedProxies();
    await testMixedWorkload();

    const totalElapsed = performance.now() - totalStart;
    console.log(`All isolated connection tests completed! Total time: ${totalElapsed.toFixed(1)}ms`);
  } catch (error) {
    console.error('Test failed:', error);
    process.exit(1);
  }
}

main().catch(console.error);

File: python/examples/isolated_connections.py
import asyncio
import time
from collections.abc import AsyncGenerator
from datetime import datetime
from typing import Any

import uvloop

from camera_ui_rpc import RPCClass, create_rpc_client


@RPCClass
class TestService:
    async def ping(self) -> str:
        return f"pong at {datetime.now().isoformat()}"

    async def heavy_computation(self, seconds: int) -> int:
        print(f"[Service] Starting heavy computation for {seconds}s")
        start = time.perf_counter()
        while time.perf_counter() - start < seconds:
            await asyncio.sleep(0.001)
        print("[Service] Heavy computation completed")
        return int((time.perf_counter() - start) * 1000)

    async def generate_data_stream(self) -> AsyncGenerator[str, None]:
        for i in range(100):
            yield f"Data chunk {i} at {datetime.now().isoformat()}"
            await asyncio.sleep(0.01)


async def test_isolated_channels():
    print("Testing Isolated Channels\n")

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-client",
        }
    )

    await client.connect()

    regular_channel = await client.channel("regular-channel")

    isolated_channel = await client.channel("isolated-channel", isolated_connection=True)

    def on_regular_message(data: Any):
        display_data = {k: v for k, v in data.items() if k != "data"}
        if "data" in data:
            display_data["data"] = f"<{len(data['data'])} bytes>"

        print(f"[Regular Channel]: {display_data}")

    def on_isolated_message(data: Any):
        display_data = {k: v for k, v in data.items() if k != "data"}
        if "data" in data:
            display_data["data"] = f"<{len(data['data'])} bytes>"

        print(f"[Isolated Channel]: {display_data}")

    regular_channel.on("message", on_regular_message)
    isolated_channel.on("message", on_isolated_message)

    print("Sending heavy traffic on isolated channel...")
    isolated_start = time.perf_counter()
    promises: list[asyncio.Task[Any]] = []
    test_buffer = b"x" * 10000  # Pre-allocate 10KB buffer
    for i in range(1000):
        promises.append(
            asyncio.create_task(
                isolated_channel.send(
                    {
                        "type": "bulk",
                        "index": i,
                        "data": test_buffer,
                    }
                )
            )
        )

    print("Testing regular channel responsiveness...")
    regular_start = time.perf_counter()
    await regular_channel.send({"type": "ping", "timestamp": time.perf_counter()})
    regular_time = (time.perf_counter() - regular_start) * 1000
    print(f"Regular channel responded in {regular_time:.1f}ms")

    await asyncio.gather(*promises)
    isolated_time = (time.perf_counter() - isolated_start) * 1000
    throughput = (1000 * 10) / (isolated_time / 1000)  # 10KB * 1000 messages
    print(f"Heavy traffic completed in {isolated_time:.1f}ms ({throughput / 1024:.1f} MB/s)\n")

    await regular_channel.close()
    await isolated_channel.close()
    await client.disconnect()


async def test_isolated_proxies():
    print("Testing Isolated Proxies\n")

    server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-server",
        }
    )

    await server.connect()
    unsub = await server.register_handler("test", TestService())

    # Setup client
    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-client",
        }
    )

    await client.connect()

    regular_proxy = client.create_proxy("test")

    isolated_proxy = client.create_proxy("test", isolated_connection=True)

    print("Starting heavy computation on isolated proxy...")
    heavy_start = time.perf_counter()
    heavy_promise = asyncio.create_task(isolated_proxy.proxy.heavy_computation(3))

    print("Testing regular proxy responsiveness...")
    ping_times: list[float] = []
    for i in range(5):
        ping_start = time.perf_counter()
        result = await regular_proxy.ping()
        ping_time = (time.perf_counter() - ping_start) * 1000
        ping_times.append(ping_time)
        print(f"Regular proxy ping {i + 1}: {ping_time:.1f}ms - {result}")
        await asyncio.sleep(0.5)
    avg_ping_time = sum(ping_times) / len(ping_times)
    print(f"Average ping time during heavy computation: {avg_ping_time:.1f}ms")

    computation_time = await heavy_promise
    total_heavy_time = (time.perf_counter() - heavy_start) * 1000
    print(f"Heavy computation completed in {computation_time}ms (total: {total_heavy_time:.1f}ms)\n")

    print("Testing streaming on isolated proxy...")
    stream_start = time.perf_counter()
    count = 0
    async for _ in isolated_proxy.proxy.generate_data_stream():
        count += 1
        if count % 20 == 0:
            print(f"Received {count} chunks from isolated stream")
    stream_time = (time.perf_counter() - stream_start) * 1000
    print(
        f"Stream completed: {count} total chunks in {stream_time:.1f}ms ({count * 1000 / stream_time:.1f} chunks/s)\n"
    )

    await unsub()

    await isolated_proxy.close()
    await client.disconnect()
    await server.disconnect()

    print("Isolated proxies test completed\n")


async def test_mixed_workload():
    print("Testing Mixed Workload\n")

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "mixed-client",
        }
    )

    await client.connect()

    chat_channel = await client.private_channel("chat", "partner", isolated_connection=True)

    data_channel = await client.private_channel("data-transfer", "partner", isolated_connection=True)

    print("Starting mixed workload...")

    chat_task_running = True

    async def send_chat_messages():
        while chat_task_running:
            await chat_channel.send(
                {
                    "type": "chat",
                    "message": "Hello from chat",
                    "timestamp": datetime.now().isoformat(),
                }
            )
            await asyncio.sleep(1)

    chat_task = asyncio.create_task(send_chat_messages())

    data_start = time.perf_counter()
    data_promises: list[asyncio.Task[Any]] = []
    data_buffer = b"x" * (100 * 1024)  # Pre-allocate 100KB buffer
    for i in range(100):
        data_promises.append(
            asyncio.create_task(
                data_channel.send(
                    {
                        "type": "data",
                        "chunk": i,
                        "payload": data_buffer,
                    }
                )
            )
        )

    chat_messages = 0
    data_chunks = 0

    def on_chat_message(_: Any):
        nonlocal chat_messages
        chat_messages += 1

    def on_data_message(_: Any):
        nonlocal data_chunks
        data_chunks += 1

    chat_channel.on("message", on_chat_message)
    data_channel.on("message", on_data_message)

    await asyncio.gather(*data_promises)
    chat_task_running = False
    await chat_task
    data_time = (time.perf_counter() - data_start) * 1000
    data_throughput = (100 * 100) / (data_time / 1000)  # 100KB * 100 messages

    print(f"Chat messages: {chat_messages}")
    print(f"Data chunks: {data_chunks}")
    print(f"Data transfer: 10MB in {data_time:.1f}ms ({data_throughput / 1024:.1f} MB/s)")
    print("Mixed workload completed\n")

    await chat_channel.close()
    await data_channel.close()
    await client.disconnect()


async def main():
    total_start = time.perf_counter()
    try:
        await test_isolated_channels()
        await test_isolated_proxies()
        await test_mixed_workload()

        total_elapsed = (time.perf_counter() - total_start) * 1000
        print(f"All isolated connection tests completed! Total time: {total_elapsed:.1f}ms")
    except Exception as error:
        print(f"Test failed: {error}")
        import traceback

        traceback.print_exc()


if __name__ == "__main__":
    try:
        uvloop.run(main())
    except KeyboardInterrupt:
        print("\nTest interrupted by user")
    except Exception as e:
        print(f"Error in native channel request test: {e}")
        import traceback

        traceback.print_exc()
